Cubical/Experiments/Brunerie.agda

```diff
@@ -0,0 +1,315 @@
+        ; (k = i0) → ∣ surf i j ∣₂
+        ; (k = i1) → ∣ surf i j ∣₂
+        ; (l = i0) → ∣ surf i j ∣₂
+        })
+     ∣ surf i j ∣₂
+
+multTwoTildeAux : (t : ∥ S² ∥₂) → Path (Path ∥ S² ∥₂ t t) refl refl
+multTwoTildeAux ∣ x ∣₂ = multTwoAux x
+multTwoTildeAux (squash₂ _ _ _ _ _ _ t u k l m n) i j =
+  squash₂ _ _ _ _ _ _
+    (λ k l m → multTwoTildeAux (t k l m) i j)
+    (λ k l m → multTwoTildeAux (u k l m) i j)
+    k l m n
+
+multTwoEquivAux : Path (Path (∥ S² ∥₂ ≃ ∥ S² ∥₂) (idEquiv _) (idEquiv _)) refl refl
+multTwoEquivAux i j =
+  ( f i j
+  , hcomp
+      (λ l → λ
+        { (i = i0) → isPropIsEquiv _ (idIsEquiv _) (idIsEquiv _) l
+        ; (i = i1) → isPropIsEquiv _ (idIsEquiv _) (idIsEquiv _) l
+        ; (j = i0) → isPropIsEquiv _ (idIsEquiv _) (idIsEquiv _) l
+        ; (j = i1) →
+          isPropIsEquiv _
+            (transp (λ k → isEquiv (f i k)) (i ∨ ~ i) (idIsEquiv _))
+            (idIsEquiv _)
+            l
+        })
+      (transp (λ k → isEquiv (f i (j ∧ k))) (i ∨ ~ i ∨ ~ j) (idIsEquiv _))
+  )
+  where
+  f : I → I → ∥ S² ∥₂ → ∥ S² ∥₂
+  f i j t = multTwoTildeAux t i j
+
+tHopf³ : S³ → Set
+tHopf³ base = ∥ S² ∥₂
+tHopf³ (surf i j k) =
+  Glue ∥ S² ∥₂
+    (λ { (i = i0) → (∥ S² ∥₂ , idEquiv _)
+       ; (i = i1) → (∥ S² ∥₂ , idEquiv _)
+       ; (j = i0) → (∥ S² ∥₂ , idEquiv _)
+       ; (j = i1) → (∥ S² ∥₂ , idEquiv _)
+       ; (k = i0) → (∥ S² ∥₂ , multTwoEquivAux i j)
+       ; (k = i1) → (∥ S² ∥₂ , idEquiv _)
+       })
+
+π₃S³ : Ω³ ptS³ .fst → Ω² pt∥ ptS² ∥₂ .fst
+π₃S³ p i j = transp (λ k → tHopf³ (p j k i)) i0 ∣ base ∣₂
+
+postulate
+  is2GroupoidGROUPOID : ∀ {ℓ} → is2Groupoid (GROUPOID ℓ)
+  isGroupoidSET : ∀ {ℓ} → isGroupoid (SET ℓ)
+
+codeS² : S² → GROUPOID _
+codeS² s = ∥ HopfS² s ∥₁ , squash₁
+
+codeTruncS² : ∥ S² ∥₂ → GROUPOID _
+codeTruncS² = rec2GroupoidTrunc is2GroupoidGROUPOID codeS²
+
+encodeTruncS² : Ω pt∥ ptS² ∥₂ .fst → ∥ S¹ ∥₁
+encodeTruncS² p = transp (λ i → codeTruncS² (p i) .fst) i0 ∣ base ∣₁
+
+codeS¹ : S¹ → SET _
+codeS¹ s = ∥ helix s ∥₀ , squash₀
+
+codeTruncS¹ : ∥ S¹ ∥₁ → SET _
+codeTruncS¹ = recGroupoidTrunc isGroupoidSET codeS¹
+
+encodeTruncS¹ : Ω pt∥ ptS¹ ∥₁ .fst → ∥ Int ∥₀
+encodeTruncS¹ p = transp (λ i → codeTruncS¹ (p i) .fst) i0 ∣ pos zero ∣₀
+
+
+-- THE BIG GAME
+
+f3 : Ω³ ptS³ .fst → Ω³ (ptjoin ptS¹ S¹) .fst
+f3 = mapΩ³refl S³→joinS¹S¹
+
+f4 : Ω³ (ptjoin ptS¹ S¹) .fst → Ω³ ptS² .fst
+f4 = mapΩ³refl alpha
+
+f5 : Ω³ ptS² .fst → Ω³ (ptjoin ptS¹ S¹) .fst
+f5 = h
+
+f6 : Ω³ (ptjoin ptS¹ S¹) .fst → Ω³ ptS³ .fst
+f6 = mapΩ³refl joinS¹S¹→S³
+
+f7 : Ω³ ptS³ .fst → Ω² pt∥ ptS² ∥₂ .fst
+f7 = π₃S³
+
+g8 : Ω² pt∥ ptS² ∥₂ .fst → Ω pt∥ ptS¹ ∥₁ .fst
+g8 = mapΩrefl encodeTruncS²
+
+g9 : Ω pt∥ ptS¹ ∥₁ .fst → ∥ Int ∥₀
+g9 = encodeTruncS¹
+
+g10 : ∥ Int ∥₀ → Int
+g10 = elimSetTrunc (λ _ → isSetInt) (idfun Int)
+
+-- don't run me
+brunerie : Int
+brunerie = g10 (g9 (g8 (f7 (f6 (f5 (f4 (f3 (λ i j k → surf i j k))))))))
+
+-- simpler tests
+
+test63 : ℕ → Int
```

Comment:
@Saizan  This function is very fast in cubicaltt, but in Cubical Agda `test63 0` terminates fast but then it gets extremely slow directly. Any ideas?

I think that being able to disable "endpoint corrections" in trans for some HITs would help a lot. These corrections are in particular not necessary for truncations, but I believe that Cubical Agda still does them?

Reply:
Yeah, Cubical Agda does the corrections for any parametrized type. I suppose we could either make Agda detect when they can be skipped, or figure out a way to provide custom implementations of transp.

Reply:
By the way, what do you mean by

> ```test63 0``` terminates fast but then it gets extremely slow directly

?

What gets slow?

Reply:
> By the way, what do you mean by
> 
> > `test63 0` terminates fast but then it gets extremely slow directly
> 
> ?
> 
> What gets slow?

He's talking about `test63 1` `test63 2` `test63 3`... etc. those computes fast in cubicaltt, but not in Cubical Agda.

Reply:
In cubicaltt the timings are as follows:
```
> test631
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m0.005s
> test632
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m0.037s
> test633
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m0.095s
> test634
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m0.209s
> test635
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m0.504s
> test636
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m1.125s
> test637
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m2.103s
> test638
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m6.096s
> test639
EVAL: pos (suc zero)
#hcomps: 0
Time: 0m9.633s
```

How can I get similar timings when running `C-c C-n test63 N` in Agda again?

Reply:
I don't know what @ecavallo was doing when we were testing this before on his laptop, but it seems to be instant to run for example `C-c C-n test63 100` on my machine?!

Reply:
Oh wait, `test63` is not testing the right thing! The base case should be `surf i j k`, not base!

Reply:
I pushed a fix now. Running `C-c C-n test63 1` does not terminate on my computer before using up all of my memory. In cubicaltt it's instant.

Reply:
@mortberg do you have some general schema on how to skip the corrections? e.g. for truncations it's not enough to remove the other ```hcomp``` you also need to skip those in the path arguments

Reply:
I'm not sure if you're asking for a syntactic criteria for when the endpoint corrections could be skipped or just how to skip them? 

For the first I don't know, maybe the best is to add a pragma to let experts flag the HIT definition themselves for now? This would let us test if this is the problem here. Alternatively we could have a way to disable them globally just so that we can test this example.

For how to implement not doing any corrections see: https://github.com/mortberg/cubicaltt/blob/pi4s3_nobug/Eval.hs#L664